include/trieste/intrusive_ptr.h

```diff
@@ -0,0 +1,341 @@
+#include <cassert>
+#include <cstddef>
+#include <functional>
+#include <ostream>
+#include <utility>
+
+namespace trieste
+{
+  namespace detail
+  {
+    // In principle, std::atomic should not be copied.
+    // It should be a single object that is pointer-to and manipulated by
+    // multiple threads. For refcounts however, it should be possible to copy a
+    // refcounted object. The catch is that _everything but the refcount should
+    // be copied_. The copy constructors here will just set the new refcount to
+    // 0, as if the object was constructed from scratch, so different
+    // intrusive_ptr can take ownership of the new object.
+    struct copyable_refcount final
+    {
+    private:
+      std::atomic<size_t> value = 0;
+
+    public:
+      constexpr copyable_refcount(size_t value_) : value{value_} {}
+
+      constexpr copyable_refcount() = default;
+      constexpr copyable_refcount(const copyable_refcount&) : value{0} {}
+      constexpr copyable_refcount(copyable_refcount&&) : value{0} {}
+
+      operator size_t() const
+      {
+        return value;
+      }
+
+      copyable_refcount& operator+=(size_t inc)
+      {
+        value += inc;
+        return *this;
+      }
+
+      size_t fetch_sub(size_t dec)
+      {
+        return value.fetch_sub(dec);
+      }
+    };
+  }
+
+  template<typename T>
+  struct intrusive_refcounted_traits
+  {
+    static constexpr void intrusive_inc_ref(T* ptr)
+    {
+      ptr->intrusive_inc_ref();
+    }
+
+    static constexpr void intrusive_dec_ref(T* ptr)
+    {
+      ptr->intrusive_dec_ref();
+    }
+  };
+
+  template<typename T>
+  struct intrusive_ptr final
+  {
+  private:
+    T* ptr;
+
+    constexpr void inc_ref()
+    {
+      if (ptr)
+      {
+        intrusive_refcounted_traits<T>::intrusive_inc_ref(ptr);
+      }
+    }
+
+    constexpr void dec_ref()
+    {
+      if (ptr)
+      {
+        intrusive_refcounted_traits<T>::intrusive_dec_ref(ptr);
+        ptr = nullptr;
+      }
+    }
+
+  public:
+    constexpr intrusive_ptr() : ptr{nullptr} {}
+
+    constexpr intrusive_ptr(std::nullptr_t) : ptr{nullptr} {}
+
+    constexpr explicit intrusive_ptr(T* ptr_) : ptr{ptr_}
+    {
+      inc_ref();
+    }
+
+    template<typename U>
+    constexpr intrusive_ptr(const intrusive_ptr<U>& other) : ptr{other.ptr}
+    {
+      inc_ref();
+    }
+
+    template<typename U>
+    constexpr intrusive_ptr(intrusive_ptr<U>&& other) : ptr{other.release()}
+    {}
+
+    constexpr intrusive_ptr(const intrusive_ptr<T>& other) : ptr{other.ptr}
+    {
+      inc_ref();
+    }
+
+    constexpr intrusive_ptr(intrusive_ptr<T>&& other) : ptr{other.release()} {}
+
+    constexpr intrusive_ptr<T>& operator=(const intrusive_ptr<T>& other)
+    {
+      // Self-assignment case (skip the code below that would destroy the object
+      // otherwise)
+      if (ptr == other.ptr)
+      {
+        return *this;
+      }
+
+      dec_ref();
+      ptr = other.ptr;
+      inc_ref();
```

Comment:
I think this code might be wrong.  The `other` is effectively borrowed as you are using a reference.  So I think it is possible for the call to decref to run a destructor, which in turn deallocates the underlying object that `other` is borrowed from.

You really should do
```suggestion
      T* tmp = ptr;
      other.ptr.intrusive_inc_ref();
      ptr = other.ptr;
      tmp.intrusive_dec_ref();
```

This should cover the self-assignment case.  You might need null-checks too.

Reply:
... yes, I think I see it.

All you need is (in some weird syntax that mixes up C++, Scala, and probably Go, sorry):
```
let foo = X
let bar = Cons(foo, Nil)

foo := Nil // bar indirectly has only ref to what foo had
bar := bar.head // destroys Cons(...) -> destroys X -> bar now points to bad
```

Reply:
Actually my previous version was basically right then, in that what you're suggesting here is equivalent up to some syntax.
The complaint was that it was doing a redundant inc_ref on self-assign, and I introduced this bug when removing the self-assign handling.

So, I put my existing solution back but with the self-assign "optimization" still there, since it also gets all the null checks right by construction (rather than give me more pointer manipulation to shoot myself in the foot with).

(working on the other comment, then committing)